File: Data/Python/dijkstra.py
import heapq

def dijkstra(graph, start):
    """
    Dijkstra's Algorithm:
    - Finds shortest path from start to all other nodes.
    - Uses a priority queue (min-heap) to get closest unvisited node.
    - Time Complexity: O((V + E) log V)
    """
    distances = {node: float('inf') for node in graph}  # Initialize distances
    distances[start] = 0  # Distance to start is 0

    min_heap = [(0, start)]  # Heap of (distance, node)

    while min_heap:
        current_dist, current_node = heapq.heappop(min_heap)

        # If current distance is already worse, skip
        if current_dist > distances[current_node]:
            continue

        for neighbor, weight in graph[current_node]:
            distance = current_dist + weight  # Total distance to neighbor
            if distance < distances[neighbor]:
                distances[neighbor] = distance  # Update shortest distance
                heapq.heappush(min_heap, (distance, neighbor))

    return distances

# ✅ Example weighted graph
weighted_graph = {
    'A': [('B', 1), ('C', 4)],
    'B': [('C', 2), ('D', 5)],
    'C': [('D', 1)],
    'D': []
}
print("Dijkstra Distances:", dijkstra(weighted_graph, 'A'))
# Output: {'A': 0, 'B': 1, 'C': 3, 'D': 4}

File: Data/C++/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
typedef pair<int, int> pii;

void dijkstra(int src, const vector<vector<pii>>& adj, int n) {
    vector<int> dist(n, INT_MAX);   // Distance from src
    priority_queue<pii, vector<pii>, greater<pii>> pq;  // Min-heap

    dist[src] = 0;
    pq.push({0, src});

    while (!pq.empty()) {
        int cost = pq.top().first;
        int u = pq.top().second;
        pq.pop();

        for (auto& [v, w] : adj[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }

    // Print shortest distances
    cout << "Shortest distances from node " << src << ":\n";
    for (int i = 0; i < n; i++) {
        cout << "To node " << i << ": " << dist[i] << endl;
    }
}

// ✅ Use-case
int main() {
    int n = 5;
    vector<vector<pii>> adj(n);
    // Add edges: u --(w)--> v
    adj[0].push_back({1, 2});
    adj[0].push_back({2, 4});
    adj[1].push_back({2, 1});
    adj[1].push_back({3, 7});
    adj[2].push_back({4, 3});
    adj[3].push_back({4, 1});

    dijkstra(0, adj, n);
}
